core/jobs_test.py

```diff
@@ -949,6 +1110,95 @@ def test_events_coming_in_while_batch_job_is_running(self):
+            with self.assertRaisesRegexp(
+                Exception,
+                'Attempted to start computation StartExplorationEventCounter, '
+                'which is already running'):
+                StartExplorationEventCounter.start_computation()
+
+            self.process_and_flush_pending_tasks()
+            StartExplorationEventCounter.stop_computation('admin_user_id')
+
+    def test_get_continuous_computations_info_with_existing_model(self):
+        job_models.ContinuousComputationModel(
+            id='StartExplorationEventCounter').put()
+        continuous_computations_data = jobs.get_continuous_computations_info(
+            self.ALL_CC_MANAGERS_FOR_TESTS)
+
+        self.assertEqual(len(continuous_computations_data), 1)
+        continuous_computation_data = continuous_computations_data[0]
+
+        self.assertEqual(
+            continuous_computation_data['computation_type'],
+            'StartExplorationEventCounter')
+        self.assertEqual(
+            continuous_computation_data['status_code'], 'idle')
+        self.assertTrue(continuous_computation_data['is_startable'])
+        self.assertFalse(continuous_computation_data['is_stoppable'])
+
+    def test_failing_continuous_job(self):
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *args):
+            """Mocks logging.error()."""
+            observed_log_messages.append(msg % args)
+
+        StartExplorationEventCounter.start_computation()
+
+        status = StartExplorationEventCounter.get_status_code()
+        self.assertEqual(
+            status, job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_RUNNING)
+
+        with self.swap(logging, 'error', _mock_logging_function):
+            StartExplorationEventCounter.on_batch_job_failure()
+
+        self.process_and_flush_pending_tasks()
+        StartExplorationEventCounter.stop_computation('admin_user_id')
+
+        self.assertEqual(
+            observed_log_messages, ['Job StartExplorationEventCounter failed.'])
+
+    def test_cancelling_continuous_job(self):
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *args):
+            """Mocks logging.error()."""
+            observed_log_messages.append(msg % args)
+
+        StartExplorationEventCounter.start_computation()
+
+        status = StartExplorationEventCounter.get_status_code()
+        self.assertEqual(
+            status, job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_RUNNING)
+
+        with self.swap(logging, 'info', _mock_logging_function):
+            StartExplorationEventCounter.on_batch_job_canceled()
+
+        self.process_and_flush_pending_tasks()
+        StartExplorationEventCounter.stop_computation('admin_user_id')
+
+        self.assertEqual(
+            observed_log_messages,
+            ['Job StartExplorationEventCounter canceled.'])
+
+    @patch('core.jobs.BaseContinuousComputationManager._kickoff_batch_job')
```

Comment:
I'm not convinced we need to introduce a brand new dependency just to test this, isn't there a simpler construct we can use? Please explain why or why not.

Reply:
"Mock" tests that _kickoff_batch_job_after_previous_one_ends calls _kickoff_batch_job. I thought since we are gonna introduce mock in milestone 3 to test the python scripts(hopefully), why not introduce it now? Also if I created a job which does not re-implement _kickoff_batch_job_after_previous_one_ends, the job will run indefinitely

Reply:
What about something like

```python
def test_method(self):
    times_run = 0
    class MockContinuousComputationManager(BaseContinuousComputationManager):
        def _kickoff_batch_job_after_previous_one_ends(self):
            if times_run < 2:
                super(self, MockContinuousComputationManager)._kickoff_batch_job_after_previous_one_ends()
                times_run += 1
    # start job
    self.assertEqual(times_run, 2)
```

Reply:
I get the same output as mentioned in the devlogs:
`TypeError: 'NoneType' object has no attribute '__getitem__'`

Reply:
Try again with our new-found knowledge about starting vs flushing tasks?

Reply:
done